### ISSUE TITLE
Breaking change inside v7.0.200 - MSBuild Preview version used !?

### ISSUE THREAD

**Issue by WernerMairl** [dotnet/sdk#30624]
### Describe the bug
"dotnet build -o  <folder>" feature (output folder for entire solution build)

![image](https://user-images.githubusercontent.com/10084630/218982850-a374375e-26e0-4bc3-8416-4fe5b5c6bb22.png)

### To Reproduce

 use "dotnet build -o <folder>" with sdk 7.0.200 from today

### Exceptions (if any)
error NETSDK1194: The "--output" option isn't supported when building a solution.

### Further technical details
You can see in the screenshot that a preview version of MSBuild is used/delivered (7.5.0-preview....".
Maybe this is the root cause of the issue!
Because with the SDK from last week (7.0.100) i can workaround the exception!

I hope that helps - assuming this issue may cause a lot of problems...

BR
Werner

**Comment by WernerMairl** [comment]
@KalleOlaviNiemitalo 
Nice fix. but following semantic versioning policy, this may be the wrong delivery :-(

**Comment by PeterLehmann** [comment]
still a breaking change you have not write in the release note

**Comment by dave-yotta** [comment]
No Semver? That's crazy! Do I have to avoid security patches just so that I can avoid breaking the build?

**Comment by WernerMairl** [comment]
all in all... 
not a good idea to deliver this in this way....

we will see how many build pipelines worldwide are broken...

🍿 🍿  🍿

**Comment by dave-yotta** [comment]
So if my runtime is `dotnet/aspnet:7.0` I need to somehow mirror that in an sdk version `dotnet/sdk:7.0.1**`?? There are no valid version strings for this in any supported tools I can think of!

**Comment by WernerMairl** [comment]
In the same time Lufthansa is grounded for IT problems...

🍿 🍹

**Comment by GregorLamche** [comment]
> So if my runtime is `dotnet/aspnet:7.0` I need to somehow mirror that in an sdk version `dotnet/sdk:7.0.1**`?? There are no valid version strings for this supported in any tools I can think of!

One reason why my team stopped using Microsoft docker images (MS is too slow to role out security updates of the base image).

As a work around you can switch to a OS base image, and install the sdk/runtime using its package manager.
With Ubuntu it would look like this:

```
FROM ubuntu:22.04 as base
RUN apt-get update \
    && apt-get install -y dotnet-sdk-7.0=7.0.102-1
```

**Comment by dave-yotta** [comment]
No, it seems it is the install-scripts. I couldn't repro directly on sdk:7.0.  

Here's what leads to the error for me:

```
FROM mcr.microsoft.com/dotnet/aspnet:7.0-bullseye-slim as base
RUN apt-get update 
RUN apt install -y jq moreutils curl
RUN curl -sSL https://dot.net/v1/dotnet-install.sh > dotnet-install.sh
RUN chmod u+x dotnet-install.sh
RUN ./dotnet-install.sh --channel 7.0 --install-dir /usr/share/dotnet

FROM base
RUN dotnet pack -c release -o packages-out
```

**Comment by WernerMairl** [comment]
In the meantime I'm working on some "forward" fix for my projects.

### UseCase
build + pack + push (Github-Action)

### Old version: 

Option "-o" used to locate the resulting *.nupkg files for push:

![image](https://user-images.githubusercontent.com/10084630/219042204-99b6f8e7-d5b1-4e31-80d2-dd37eeaf62c5.png)

### New version 

for SDK 7.0.200:
- removed all the "-o" usages
- modified dotnet nuget push for recursive folder search:

![image](https://user-images.githubusercontent.com/10084630/219041291-237a3574-ffce-499f-aca3-82b4b7f47cd0.png)

**Comment by MichaelSimons** [comment]
> > So if my runtime is `dotnet/aspnet:7.0` I need to somehow mirror that in an sdk version `dotnet/sdk:7.0.1**`?? There are no valid version strings for this supported in any tools I can think of!
> 
> One reason why my team stopped using Microsoft docker images (MS is too slow to role out security updates of the base image).
> 
> As a work around you can switch to a OS base image, and install the sdk/runtime using its package manager. With Ubuntu it would look like this:
> 
> ```
> FROM ubuntu:22.04 as base
> RUN apt-get update \
>     && apt-get install -y dotnet-sdk-7.0=7.0.102-1
> ```

I don't want to derail this topic here, @GregorLamche, would you care to start a [discussion topic](https://github.com/dotnet/dotnet-docker/discussions) on your experience?  I'd like to know details and try to improve our .NET Docker offering if we are not meeting your needs.  TIA.

**Comment by amis92** [comment]
It's very disturbing to see `-preview` label when running a simple `dotnet build` using the this SDK.

```
$ dotnet build --no-restore
MSBuild version 17.5.0-preview-23061-01+040e2a90e for .NET
```

And, I was also hit by the breaking change, quite involutarily, as my global.json says 7.0.100 but `actions/setup-dotnet` installed the new one anyway.

**Comment by baronfel** [comment]
We've been discussing the introduction of the new error and will be taking the following actions based on everyone's feedback in this (and other!) threads - THANK YOU for making your feedback known, please keep it up in the future.

a) We'll downgrade the error to a warning - we think this usage pattern _is_ harmful and can lead to builds that are harder to debug, especially when projects are multi-targeted or have differing sets of PackageReference versions.
b) We'll skip the logic for the `pack` command entirely - `pack` has stable and correct semantics when the `-o` flag is specified (because it sets PackageOutputPath instead of `OutputPath` directly)

We will put this up for inclusion into next month's servicing release of the .NET SDK, which is the soonest possible ship vehicle we have for this kind of change. In the meantime, you can restore the previous behavior by changing the use of `-o` to a specific property. If you want to maintain the existing behavior exactly, then you can use the `--property` flag to set a MSBuild property to the desired directory. The property to use varies based on the command:

| Command | Property | Example |
| -- | -- | -- |
| `build` | `OutputPath` | `dotnet build --property OutputPath=DESIRED_PATH` |
| `clean` | `OutputPath` | `dotnet clean --property OutputPath=DESIRED_PATH` |
| `pack` | `PackageOutputPath` | `dotnet pack --property PackageOutputPath=DESIRED_PATH` |
| `publish` | `PublishDir` | `dotnet publish --property PublishDir=DESIRED_PATH` |
| `store` | `OutputPath` | `dotnet store --property OutputPath=DESIRED_PATH` |
| `test` | `TestResultsDirectory` | `dotnet test --property OutputPath=DESIRED_PATH` |

**Comment by ernestopye** [comment]
We just hit this today as well. It appears to be an issue with 6.0.406 too (using MSBuild 17.5.0-preview-23061-01+040e2a90e). We were targeting 6.0.x and went to this version after it was released automatically. 

We've now pinned our version to 6.0.405 and we're back to using a non-prerelease version of MSBuild on our build agents.

We have dozens of pipelines that will be affected by this and will need to update all of them. 

We had no errors at all. It was just that ViewComponents were simply not compiling, and that's how it was brought to our attention.

**Comment by Dreamescaper** [comment]
@baronfel 
Unfortunately, setting PackageOutputPath parameter didn't help with relative path "./packages". Instead of putting all the packages in the same folder, it has created separate "packages" folders inside of every project directory.

I've added smth like that to workaround that:
```
        $path = [System.IO.Path]::GetFullPath("packages")
        dotnet pack --property PackageOutputPath=$path -c Release --no-restore
```

**Comment by baronfel** [comment]
Can you set the path using an absolute path instead?

**Comment by Dreamescaper** [comment]
Not really, because it is executed from GitHub Actions, I don't know the absolute path upfront.

But that's what I essentially did with this line:
> $path = [System.IO.Path]::GetFullPath("packages")

**Comment by GregorLamche** [comment]
@MichaelSimons I don't think there would be much gain for you in me doing this. We have a rather unique setup which wouldn't be much help to many other people. DotNet, and ASP.Net are just two of many tools we use to achieve our goals. It's simply that having many individual parts that we can control individually is easier than one external base image that tries to catch it all.

Frankensteining a service together sometimes just is the best option.

**Comment by GregorLamche** [comment]
> Command 	Property 	Example
> `build` 	`OutputPath` 	`dotnet build --property OutputPath=DESIRED_PATH`
> `clean` 	`OutputPath` 	`dotnet clean --property OutputPath=DESIRED_PATH`
> `pack` 	`PackageOutputPath` 	`dotnet pack --property PackageOutputPath=DESIRED_PATH`
> `publish` 	`PublishDir` 	`dotnet publish --property PublishDir=DESIRED_PATH`
> `store` 	`OutputPath` 	`dotnet store --property OutputPath=DESIRED_PATH`
> `test` 	`TestResultsDirectory` 	`dotnet test --property OutputPath=DESIRED_PATH`

I just updated our Dockerfile, and I gotta say, I'm not a fan of the new approaches you've mentioned. For one, the mentioned flags don't work for all commands and written. Here's how it works on my end (truncated):

```
dotnet build service.sln --property OutputPath=/app/build 
dotnet test service.sln /p:OutputPath=/app/build //The --property approach returned "MSB1005: Specify a property and its value.", probably because of the Docker RUN [] that calles it.
dotnet publish src/WebService/WebService.csproj --property PublishDir=/app/publish
```

I think it's rather unoptimal that what is pretty much the same functionality needs to be addressed through 3 different flags and arguments. And that's only the 3 commands that I need.

**Comment by bradr-stadion** [comment]
This .NET SDK version completely broke Razor compilations, and caused our developers to lose 2 days worth of development time.
Why was the latest SDK download task able to pull down a preview version?! 
Specifying the version number in the download task to the last major release resolved the issue.

**Comment by bradr-stadion** [comment]
> > Can you please detail what you mean about "Specifying the version number in the download task to the last major release resolved the issue."? I've tried the other work arounds and none work for dotnet publish.

Hi Shaun,

![image](https://user-images.githubusercontent.com/119293212/219411793-72d6ddbd-261e-4329-b4f8-f9649141d826.png)

In our build task, we specified the version number as shown above.

Hope this helps!

**Comment by baronfel** [comment]
thank you @Dreamescaper and @GregorLamche - I've updated the examples in the tables to be universally correct - it seems that the `test` command isn't using the nicer option property parsing from the .NET CLI.  I've also added a clarification that the path must be absolute for best results.

**Comment by baronfel** [comment]
The other issue reported here is the MSBuild version being a preview version - this happened because our dependency-flow tooling got slightly out of sync, so the MSBuild in the downloadable SDK is actually slightly _ahead_ of the MSBuild shipped in VS.  We will roll back the versions for next month's shipped SDK (that PR is [happening now](https://github.com/dotnet/sdk/pull/30674)). The delta between the versions can be seen [here](https://github.com/dotnet/msbuild/compare/040e2a90ee794997fc9716487280a54b0ad8fe2a...6f08c67f3ed838dccfbcdab91e6bebc54a26fd94) and contains 3 commits:

* one versioning commit
* one commit for the FileTracker, which isn't relevant to .NET (only .NET Framework)
* one commit limiting the API surface area for the not-yet-released LiveLogger feature

All of this is to say that you can rely on this version of MSBuild to behave correctly, and the versions will be corrected next month.

**Comment by ernestopye** [comment]
> All of this is to say that you can rely on this version of MSBuild to behave correctly, and the versions will be corrected next month.

There seems to be a lack of urgency on getting this fix out, so then should we consider it bad practice to allow our builds to automatically update to the latest minor version of the SDK? It sounds like we can't rely on minor version updates to the SDK to be stable, so it would follow that we need to pin the SDK version going forward, and test thoroughly after each update. Can you please confirm that that is the expectation for application developers?

**Comment by baronfel** [comment]
You should feel free to float your versions, and we shouldn't break you. That harms uptake of new versions, and encouraging folks to stay up to date is one of our primary goals as the SDK team. The feedback on excess breakage from this change is the primary motivation for walking it back!

As to lack of urgency, we (the SDK team) are limited at a process level to ship once a month, and due to various handoffs and validations this isn't likely to change soon. Next month's release is literally the earliest possible delivery window we can give you. This isn't unique for this change, either - if you look at the issue/PR history for this repo for items with the `servicing-approved` label you'll see a long history of fixes being subject to this timeline. Again, we apologize for the breakage, and it is entirely reasonable on users part to pin to a version for a month until a fixed release is out.

Now, putting on my MSBuild hat, you should really look at _why_ we introduce the error (soon to be warning) - solution-level builds with an output directory specified are ticking time-bombs, and moving to something more repeatable is a good thing to think about. We have some [designs proposed for .NET 8](https://github.com/dotnet/designs/pull/281) that make make solution-level builds less error-prone that may be of interested to folks doing solution-level builds as well.

**Comment by baronfel** [comment]
@dave-yotta that depends - when fixes happen in the runtime components of .NET (so specifically not the SDK code, CLI , etc), those are applied to every in-support feature band of the .NET tooling. This is why on Patch Tuesday this month you saw patch updates for the following SDK feature bands:

* 6.0.100
* 6.0.300
* 6.0.400
* 7.0.100
* 7.0.200 

These are in support for one of 3 reasons:
* They are the current latest feature-band for an in-support major version of the Runtime (so 6.0 or 7.0), 
* They are a feature band that is part of an in-support version of VS, 
* They are the .100-band of a major version of the Runtime and so part of our support agreement with source-build, which is used by our Linux partners to build their own packaging of .NET

For fixes to the SDK and tools (which covers MSBuild, the dotnet CLI, everything in the dotnet/SDK repo, etc) we have to get approval for each fix - this is a decision made by people that aren't me that balances the impact to end users with the risk of the change, and part of that decision is which feature bands we'll service for. When that decision is made, you'll see the `servicing-approved` label appear on the PR, the PR will be merged to the _earliest_ feature band that the problem will be fixed in, and then our automated codeflow will flow the change forward to all of the more recent fixes.

So all of that is to say a big giant It Depends. Something like this change, which was introduced in 7.0.200, will only need to be applied to the 7.0.200 feature band and so will be applied to the next 7.0.200 release, which is 7.0.201. Does all of that make sense?

**Comment by ernestopye** [comment]
@baronfel Apologies if I'm missing something, or if I didn't provide enough context.

>  and we shouldn't break you

Historically, this has been the case, yes and we've been happy with the process. Unfortunately, this version of MSBuild seems to have also completely broken ViewComponents in ASP.NET Core. For us, there was no error - it just silently built broken DLLs, which ended up getting deployed to staging environments, breaking several of our sites. Fortunately, none of them made it to production. I just wanted to expand on this bit because it's not that our builds stopped working - they went through just fine, but the artifacts were "corrupted" so to speak. See https://github.com/dotnet/razor/issues/8281

We're pinning the SDK version as mentioned for all of our projects for now, but the gravity of the issue is causing a lot of internal pushback in terms of going back to floating these minor versions once this issue is resolved.

**Comment by baronfel** [comment]
@ernestopye I want to dig in a bit more - I think you're conflating the MSBuild version (the build engine) with changes/new features made in the C# compiler that shipped with the 7.0.200 SDK. Those are two separate things, and I want to make sure that the fact that MSbuild is reporting a preview version but contains no functional changes isn't something actually impactful for your use case.

The components that are shipped as part of the SDK (like Roslyn!) can have bugs. It's going to happen, our coverage isn't and cannot be perfect. When that happens, pinning is the correct answer for you and all impacted users. We should communicate to you the timelines for a fix, and then when that fix ships you can remove the pin and continue riding that sweet upgrade train.  Minor regressions happen _quite often_ (again, check the PR history with the `servicing-approved` label for more details here) but we as an org generally try to hop on them and iterate on a fix as quickly as we can.

**Comment by dave-yotta** [comment]
> the earliest feature band that the problem will be fixed in
It does make sense - my worry is about security related fixes not reaching the band I have pinned to.  Are such fixes always backported to the earliest feature band? I would rather have a broken build than a compromised application.

**Comment by baronfel** [comment]
@dave-yotta security fixes are applied to all in-support bands, o at this exact moment that's 6.0.100, 6.0.300, 6.0.400, 7.0.100, and 7.0.200. You can check out the fine print for the support policy over at [our Support Policy page](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core).

**Comment by dave-yotta** [comment]
>These are in support for one of a few reasons, like:
> They are the current latest feature-band for an in-support major version of the Runtime (so 6.0 or 7.0),

I take this to mean that all feature bands of 7.0 will be "in-support bands" so long as .NET7 is in support?

**Comment by ernestopye** [comment]
> I want to dig in a bit more - I think you're conflating the MSBuild version (the build engine) with changes/new features made in the C# compiler that shipped with the 7.0.200 SDK.

You're right. I found this issue in the related one that I linked. That's on me. We're specifically on 6.0.x, and 6.0.406 is the SDK version that's causing problems for us. I'll see if I can find a separate issue about this to keep the discussion focused.

Unfortunately, given the number of projects we're dealing with, it can be quite challenging to quickly pin this version across many build definitions. We may need to stick to pinning going forward.

**Comment by baronfel** [comment]
> > These are in support for one of a few reasons, like:
> > They are the current latest feature-band for an in-support major version of the Runtime (so 6.0 or 7.0),
> 
> I take this to mean that all feature bands of 7.0 will be "in-support bands" so long as .NET7 is in support?

Close but not quite - check out [this doc](https://learn.microsoft.com/en-us/dotnet/core/porting/versioning-sdk-msbuild-vs) on .NET SDK and Visual Studio versioning for more details, but the 200-band of each SDK generally has a much shorter support lifetime because it's tied to a VS version with a shorter support lifecycle. It's confusing, but the general rule so far is 100/300/400 are the supported bands for each supported Long-Term or Standard-Term version of the SDK.

**Comment by boomalator** [comment]
> We will put this up for inclusion into next month's servicing release of the .NET SDK, which is the soonest possible ship vehicle we have for this kind of change. In the meantime, you can restore the previous behavior by changing the use of `-o` to a specific property. If you want to maintain the existing behavior exactly, then you can use the `--property` flag to set a MSBuild property to the desired directory. The property to use varies based on the command:

@baronfel 

Um, okay, sure....  I have to go learn something new because the Azure App Development Center produces a script that doesn't work with the current version of .Net? And the "earliest release vehicle" to fix this is _weeks_ away?

> Now, putting on my MSBuild hat, you should really look at why we introduce the error (soon to be warning) - solution-level builds with an output directory specified are ticking time-bombs, and moving to something more repeatable is a good thing to think about. We have some https://github.com/dotnet/designs/pull/281 that make make solution-level builds less error-prone that may be of interested to folks doing solution-level builds as well.

So, fine, where's the guidance/docs to NOT do this? I can't find anything that tells me how to have a GitHub action build and publish _just the project_ or whatever it is you want it to build, which appears to be something different than the Azure Tools want Build to build.

We have a really simple build process. We push a commit to a branch. A GitHub action (written and supplied by Microsoft in the Azure Deployment Center) restores, builds, publishes and updates the Web App. This broke overnight. I checked the Azure Docs and even redid the deployment in case I missed something. Has anyone on THIS team talked to anyone on THAT team to make sure that everything is updated together? (Evidence would suggest the answer is "no".)

**Comment by boomalator** [comment]
Changing the Azure-provided script from `run: dotnet publish -c Release -o ${{env.DOTNET_ROOT}}/myapp`
to `run: dotnet publish -c Release --property PublishDir ${{env.DOTNET_ROOT}}/myapp`

Resulted in this new, exciting error:
````
Run dotnet publish -c Release --property PublishDir C:\Users\runneradmin\AppData\Local\Microsoft\dotnet/myapp
  dotnet publish -c Release --property PublishDir C:\Users\runneradmin\AppData\Local\Microsoft\dotnet/myapp
  shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
  env:
    DOTNET_ROOT: C:\Users\runneradmin\AppData\Local\Microsoft\dotnet
Unhandled exception: System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Microsoft.DotNet.Cli.ReleasePropertyProjectLocator.GetGlobalPropertiesFromUserArgs(ParseResult parseResult)
   at Microsoft.DotNet.Cli.ReleasePropertyProjectLocator.GetCustomDefaultConfigurationValueIfSpecified(ParseResult parseResult, String defaultedConfigurationProperty, IEnumerable`1 slnOrProjectArgs, Option`1 configOption)+MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at Microsoft.DotNet.Tools.Publish.PublishCommand.FromParseResult(ParseResult parseResult, String msbuildPath)
   at Microsoft.DotNet.Tools.Publish.PublishCommand.Run(ParseResult parseResult)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.DotNet.Cli.Parser.<>c__DisplayClass17_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__17_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass15_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass7_0.<<UseExceptionHandler>b__0>d.MoveNext()
Error: Process completed with exit code 1.
````

**Comment by WernerMairl** [comment]
> a) We'll downgrade the error to a warning - we think this usage pattern _is_ harmful and can lead to builds that are harder to debug, especially when projects are multi-targeted or have differing sets of PackageReference versions. 

@baronfel 
any numbers/telemetry about how many of the broken build pipelines are using "**TreadWarningsAsErrors**" !?
At the end the pipeline definition must changed anyway...

just my 2cent!
- I totally agree to deliver this with 8.0
- For my personal/private 7.0 projects I do a forward fix rewriting my pipelines without "-o".
But i can image a endless number of 6.0/7.0 projects that are broken now and remaining broken with the proposed fix ...

BR 
Werner

**Comment by john681611** [comment]
While I commend the actions of getting on a Security issue quickly. 
I'm not clued up on the CVE (as it's got little info) but was there no other solution than just lopping off a CLI feature and dumping an error in its place?
The mechanism to publish to a single folder was super useful in pipelines.

**Comment by buvinghausen** [comment]
> @baronfel Unfortunately, setting PackageOutputPath parameter didn't help with relative path "./packages". Instead of putting all the packages in the same folder, it has created separate "packages" folders inside of every project directory.
> 
> I've added smth like that to workaround that:
> 
> ```
>         $path = [System.IO.Path]::GetFullPath("packages")
>         dotnet pack --property PackageOutputPath=$path -c Release --no-restore
> ```

PackageOutputPath starts where your csproj is located so if you wanted it in a packages subfolder of the root and you had your folder structure /src//{Project} then you would need to make it ../../packages then you can skip all the get full path stuff entirely.

**Comment by baronfel** [comment]
>  was there no other solution than just lopping off a CLI feature and dumping an error in its place

@john681611 you're right, this change was outside of our normal team process, which is to introduce new validations/changes as a warning initially, then tighten up to an error after at least one major version release. It's completely our fault that it shipped in this state, and users should be able to rely on that process to guide their update strategies.

> any numbers/telemetry about how many of the broken build pipelines are using "TreatWarningsAsErrors"

@WernerMairl we try to be [very explicit](https://learn.microsoft.com/en-us/dotnet/core/tools/telemetry#data-points) about what telemetry data we collect from the CLI, and up to this point we haven't collected data on SDK selection or settings like `TreatWarningsAsErrors` we may consider doing this, but we try not to over-collect, or collect for specific questions.

**Comment by boomalator** [comment]
> @john681611 you're right, this change was outside of our normal team process, which is to introduce new validations/changes as a warning initially, then tighten up to an error after at least one major version release. It's completely our fault that it shipped in this state, and users should be able to rely on that process to guide their update strategies.

@baronfel  Not only "users". Azure has taken a dependency on using MSBuild for it's suggested approach to CI. I haven't checked today, but as of yesterday, Azure was was creating scripts that _don't work_.   The Azure doc's are now wrong, the Azure guidance is now wrong, the Azure help desk is unaware that anything is wrong.

Aside from that, as a proficient but specific developer who follows a lot of MS templates and How-To's, I can find NOTHING that explains what I'm doing "wrong" in my building and publishing and what I'm supposed to change to "not build a solution, just a project".   And for that matter, why it's okay and supposedly "safe" to download the publish profile and click "publish" in the VS GUI, but it's so terribly dangerous to run the Microsoft-authored Azure-provide GitHub-endorsed Action (script) that there was an immediate need to halt all such publishing, overnight, without any warning.

If this issue is that big a problem, it needs to have explanations, guidance and doc updates prepared quickly as well.

**Comment by baronfel** [comment]
@boomalator you are also correct about that - I'm working on enhancements to the breaking change notifications specifically around that in the interim. Regarding "not build a solution, just a project", I mean to pass the csproj to build as the argument to dotnet build, pack, etc. I will update the comment and guidance to be more specific and/or point to the argument documentation for those commands.

**Comment by jamesbascle** [comment]
What exactly are we supposed to do about this?  This has broken hundreds of build/release pipelines.  What is the recommended replacement here?

**Comment by marcpopMSFT** [comment]
I wanted to follow up on @baronfel's comments to update everyone on our current plan. 

We are working on a 7.0.201 hotfix release. This will completely unblock all customers that don't have warnaserror enabled and provide a more standard path for disabling the warning for those who do. After internal discussions, we've decided that waiting until next month's servicing release would be too long to wait to provide relief to customers. We're hoping to release the hotfix next week, which will be the 7.0.200 build plus this change: https://github.com/dotnet/sdk/pull/30660 to move the error down to a warning.  

We will update our known issues and breaking change guidance as that option becomes available. The March servicing release of the 7.0.2xx .NET SDK will become 7.0.202. I do not believe we have done an out of band hotfix for the SDK previously but we've reviewed the process and believe it is low enough risk and worth doing. 

In responding to regular feedback from other customers who were struggling to investigate build failures caused by all outputs going to the same folder, we were overly aggressive in making the initial change. That change missed our breaking change process and we should have reviewed it with the community first, documented it months ago when it was first available in daily builds and 17.5-preview2, and gathered feedback before proceeding. We will be reviewing our breaking changes process to avoid an incident like this in the future. We apologize for the customer pain caused and hopefully the issue will be resolved next week.

https://github.com/dotnet/core/blob/main/release-notes/7.0/known-issues.md#70200-using-the---output-option-fails-for-many-commands-when-targeting-a-solution

**Comment by asottile** [comment]
are there any workarounds for github actions in the meantime?  I imagine even with a hotfix we're going to have to wait a few weeks for that to reach GA in the runner images

[pre-commit](https://github.com/pre-commit/pre-commit) uses `dotnet build` to support tools written in `dotnet` (and those language tests are now broken due to the updated versions in github actions)

**Comment by jdevine7** [comment]
workaround for azure pipelines: 
`    
- task: UseDotNet@2
      displayName: 'Use .NET Core sdk 7.0.100'
      inputs:
        version: 7.0.100
`

**Comment by peter-at-work** [comment]
> are there any workarounds for github actions in the meantime? I imagine even with a hotfix we're going to have to wait a few weeks for that to reach GA in the runner images
> 
GitHub Actions were indeed affected, even when we have specified _dotnet-quality_ `ga` on the `setup-dotnet@v3` action.
We migrated over to the recommended parameter like this (should work on both Linux shells and pwsh):
```
dotnet pack -c Release MySolution.sln -p:PackageOutputPath=$PWD/packages
```

**Comment by snwflake** [comment]
I just want to add, that the option parameter isn't the only thing wrong with this release.

https://github.com/dotnet/razor/issues/8281

We spent solid 3 hours trying to debug why none of our ViewComponents were rendering anymore. I started digging through different CI output because I was out of options at that point, then I noticed the the following line `MSBuild version 17.5.0-preview-23061-01+040e2a90e for .NET`

Rolling back the runner to 7.0.103 fixed the issue for us.

**Comment by baronfel** [comment]
@asottile workarounds include 
* pinning to a previous version (you can use the `dotnet-version` parameter set to a version like `7.0.103` to get previous behavior), 
* using a global.json file to accomplish the same (since the setup-dotnet action can use the global.json value to determine what version to use)
* using the properties described in [this comment](https://github.com/dotnet/sdk/issues/30624#issuecomment-1432118204) to adjust your CI scripts instead of the `-o`/`--output` option, or
* moving your CI actions away from building/publishing entire solutions and instead publishing/packing specific projects

The correct answer for each team will differ, but the simplest solutions in my mind would be to pin using one of the first two options. The latter two require doing more work to investigate your builds, which may be more of a time commitment to solve.

**Comment by snwflake** [comment]
@baronfel Pinning the version in actions/setup-dotnet does *not* work.

See https://github.com/actions/setup-dotnet/issues/383

**Comment by asottile** [comment]
> @asottile workarounds include
> 
>     * pinning to a previous version (you can use the `dotnet-version` parameter set to a version like `7.0.103` to get previous behavior),
> 
>     * using a global.json file to accomplish the same (since the setup-dotnet action can use the global.json value to determine what version to use)
> 
>     * using the properties described in [this comment](https://github.com/dotnet/sdk/issues/30624#issuecomment-1432118204) to adjust your CI scripts instead of the `-o`/`--output` option, or
> 
>     * moving your CI actions away from building/publishing entire solutions and instead publishing/packing specific projects
> 
> 
> The correct answer for each team will differ, but the simplest solutions in my mind would be to pin using one of the first two options. The latter two require doing more work to investigate your builds, which may be more of a time commitment to solve.

so the tricky part is `dotnet build --output` is the current "interface" that my tool uses to build dotnet things, so I'm put in a tough spot (either way I would be passing along a breaking change to my users and consumers).  I'm also not using `setup-dotnet`, just accepting whatever GHA has installed -- from what I've read `setup-dotnet` can't be overridden in this way so it feels like I'm stuck

**Comment by baronfel** [comment]
@snwflake does your project have a global.json?

**Comment by snwflake** [comment]
@baronfel yes, as per the warning outlined in the actions readme

**Comment by baronfel** [comment]
@snwflake so to be super clear, you have a global.json with a version set to `7.0.103`, and you've set `7.0.103` in your github action, and still 7.0.200 is being picked up? Can you add a call to `dotnet --info` in your pipeline and report back? That sounds unlike any scenario I've seen with pinning in the setup-dotnet action.

@asottile you're thinking of [this line](https://github.com/pre-commit/pre-commit/blob/main/pre_commit/languages/dotnet.py#L72) I imagine? In that case, I would suggest being _very_ future proof and changing from using `--output` to using `--property PackageOutputPath=<fully qualified path>` - that is virtually guaranteed to not change and to work, and in fact is what the `--output` flag was translated to in all prior versions of the .NET CLI.

**Comment by snwflake** [comment]
@baronfel disregard! Turns out, I'm an idiot

> allowPrerelease
> Type: boolean
> Available since: .NET Core 3.0 SDK.
> Indicates whether the SDK resolver should consider prerelease versions when selecting the SDK version to use.
> 
> If you don't set this value explicitly, the default value depends on whether you're running from Visual Studio:
> 
> If you're not in Visual Studio, the default value is true.

I missed that bit in the docs.

**Comment by emmyforeignet** [comment]
#30621

**Comment by neilboyd** [comment]
The instructions here also need to be updated:
https://learn.microsoft.com/en-us/azure/azure-functions/functions-how-to-github-actions?tabs=dotnet#build-the-function-app

**Comment by baronfel** [comment]
> The instructions here also need to be updated: [learn.microsoft.com/en-us/azure/azure-functions/functions-how-to-github-actions?tabs=dotnet#build-the-function-app](https://learn.microsoft.com/en-us/azure/azure-functions/functions-how-to-github-actions?tabs=dotnet#build-the-function-app)

I took at a look at this sample and I don't think it's susceptible to this change @neilboyd - the Azure Functions VSCode extension only generates a project, not a solution and a project on the same directory level, and so using the `--output` flag has well-understood semantics.

**Comment by neilboyd** [comment]
> > The instructions here also need to be updated: [learn.microsoft.com/en-us/azure/azure-functions/functions-how-to-github-actions?tabs=dotnet#build-the-function-app](https://learn.microsoft.com/en-us/azure/azure-functions/functions-how-to-github-actions?tabs=dotnet#build-the-function-app)
> 
> I took at a look at this sample and I don't think it's susceptible to this change @neilboyd - the Azure Functions VSCode extension only generates a project, not a solution and a project on the same directory level, and so using the `--output` flag has well-understood semantics.

The instructions build a solution (AZURE_FUNCTIONAPP_PACKAGE_PATH defaults to the repository root).
And that page is not about the VSCode extension.

**Comment by baronfel** [comment]
@neilboyd Can you walk me through where the source code being built by this documentation is? I was following the linked [Quickstart: Create a function in Azure using Visual Studio Code](https://learn.microsoft.com/en-us/azure/azure-functions/create-first-function-vs-code-csharp) from that same page and only got a project, so I'm missing some connection.

**Comment by xt0rted** [comment]
The issues with this release are why I like to pin to a specific sdk version, and not enable roll forward. Most people I talk to don't like this, but it helps prevent these types of random issues from popping up.

At work we enable roll forward and our builds broke as a result of `dotnet publish App.sln --no-build --output ...` now throwing an error. I know `dotnet pack` is excluded from this check now, but `dotnet publish App.sln --no-build` should be too since concurrent building was the reason for enabling this.

The workflow I use in my projects is to pin without roll forward, and then PR sdk updates to my `global.json` with GitHub Actions similar to Dependabot. This lets me treat the sdk like any other versioned dependency in my project.

- GitHub Action to handle the sdk update: https://github.com/xt0rted/dotnet-sdk-updater
- The shared workflow all my projects use: https://github.com/xt0rted/.github/blob/main/.github/workflows/dotnet-sdk-updater.yml
    - A GitHub App is used so I don't need a PAT when creating/updating the PR, but either method will work
- Example PR: https://github.com/xt0rted/dotnet-run-script/pull/150

**Comment by baronfel** [comment]
Hello folks, wanted to give you all an update on the situation.

Last night, 7.0.201 was released through all the various channels. It should already be available on GitHub Actions and AzDO runners, and of course for local installation. This hotfix contained the changes to the `--output` parameter mentioned above: it is downgraded to a warning, and the check no longer applies to the `pack` command. 

The other problem logged in this issue, MSBuild logging a prerelease version string, will be handled in the normal servicing schedule - likely with next month's expected release of 7.0.202.

Thank you all for the feedback and discussion on this issue.

**Comment by Astrophizz** [comment]
Just a heads up, 7.0.200 was pushed to our Microsoft Hosted Azure Dev Ops build agents yesterday and broke `dotnet restore --locked-mode`. The restored packages no longer adhere to our packages.lock.json.

**Comment by baronfel** [comment]
@Astrophizz do you have any details on that? maybe a binlog? paging @nkolev92 in case he's seen this reported elsewhere already.

**Comment by nkolev92** [comment]
@Astrophizz Are you using CPM? 

We haven't heard anything about lock file discrepancies yet. It'd be great if you can get us a repro and file an issue at https://github.com/nuget/home/issues

**Comment by Astrophizz** [comment]
@nkolev92 What is CPM? I'll follow up with any reproduction at the NuGet GitHub.

**Comment by nkolev92** [comment]
CPM is central package management, https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management. 
There's active development there, so I was thinking it's more likely that you're problems there thna without it. 

Looking forward to your repro!

**Comment by Astrophizz** [comment]
@nkolev92 No, we aren't using that. It sounds cool though! It occurred to me that we are using Azure pipeline caching on the NuGet package cache (keyed to a hash of the lock file) and I suppose that could introduce an extra wrinkle, though I'm not sure how or why it would fail when no changes were made to package refs between the good and bad build... 🤔

**Comment by nkolev92** [comment]
The error message, NU1004 should hopefully contain some details about what changed.

**Comment by Astrophizz** [comment]
@nkolev92 It doesn't have an error. It completes the package restore but with different packages than those specified by the lock file.

**Comment by nkolev92** [comment]
Are you running with `--locked-mode` on the CI?

**Comment by Astrophizz** [comment]
@nkolev92 Yes, I'm running it with `--locked-mode`. It appears as though that no longer enforces the lock file with 7.0.200.

**Comment by Astrophizz** [comment]
@nkolev92 After some further investigation, this looks like a matter of an Azure DevOps task restoring multiple projects ( `**/*.csproj` style with some being sibling "root" projects) and then another task building those projects. One of the root projects uses a package lock but that doesn't seem to matter much in the end as all of the built projects stomp on each other. I don't know why the result of the dll stomping changed with 7.0.200, but I suppose it could be a side effect of CPM work.

It sounds like CPM might be the solution to my problem at the same time since my reading of the docs is that I can get consistent transitive package versions across multiple solutions/root projects 😁👍. Does it support a package lock file? I'm guessing I would have a lock for each root project.

**Comment by nkolev92** [comment]
> One of the root projects uses a package lock but that doesn't seem to matter much in the end as all of the built projects stomp on each other. I don't know why the result of the dll stomping changed with 7.0.200, but I suppose it could be a side effect of CPM work

I wouldn't expect that to be the case. Most of the work in CPM is on transitive pinning, so stomping shouldn't be something that happens. 

> It sounds like CPM might be the solution to my problem at the same time since my reading of the docs is that I can get consistent transitive package versions across multiple solutions/root projects

Note that you'll only get that for transitive versions you decide to pin and if you enable transitive pinning.
Depending on how many or what kind of runnable/deployable projects you have, you may not need pinning though. 

CPM is excellent for large repos, as you'll never accidentally direct reference different versions of packages. 

> Does it support a package lock file? I'm guessing I would have a lock for each root project

Yes, they're separate features that can work together.

**Comment by baronfel** [comment]
@ericsampson the problem is how that shared `PublishDir` property is used - when set as a global property each project gets the same value for `PublishDir` and the outputs of all projects are copied to that same directory in a nondeterministic manner. This differs from `PackageOutputPath` because pack produces non-clobbering outputs to that shared path. The nondeterminism is the root of the warning here, because too many things are using that one property directly instead of namespacing or deriving a value from it.

I'd love to change the meaning of `-o` for publish to something explicitly not `PublishDir`, and then let `PublishDir` be computed based off of that root, but we need to do a bunch of user development before making any breaking changes here. I'm hopeful that approaches like the [proposed new output paths](https://github.com/dotnet/designs/blob/simplify-output-paths-2/accepted/2023/simplify-output-paths.md) will provide the kind of clean break this scenario needs, while keeping the mechanism opt-in.

**Comment by ericsampson** [comment]
thanks @baronfel.
one thing re that linked proposal, from a user perspective please don't make it enabled by default in .NET 8
It really sucks when CI breaks just by bumping .NET version, I spent a few weeks fixing that for hundreds of builds back when there was a breaking path change in the CLI behavior back in the Core 2 -> 3 transition.

I'd much much rather it to be opt-in for .NET 8, and then defaulted on for .NET 9
That allows eager people to opt in, and gives time for people to learn about the change and adjust pipelines at their leisure.

**Comment by baronfel** [comment]
@ericsampson don't worry - on by default is not in the cards currently. [Details in this section](https://github.com/dotnet/designs/blob/simplify-output-paths-2/accepted/2023/simplify-output-paths.md#changing-the-default-output-path-format), but everything will be opt in at least for .NET 8 - we recognize that there are huge compat concerns with silently changing output paths!
